File: macos/Sources/PostioKit/Reachability.swift
import Foundation
import Network

/// Watching the network path, and telling the engine about it.
///
/// **The platform observes; Rust is told.** `postio-runtime`'s watcher is
/// NetworkManager over D-Bus and returns immediately here, leaving the state
/// unknown; binding `NWPathMonitor` in Rust would mean `unsafe` in a crate
/// that forbids it, for a signal `Network` hands over in a few lines.
///
/// This does not make anything work that did not. The engine reconnects with
/// backoff on its own and runs with no signal at all — pull this out and the
/// application still syncs, only slower to notice. What it buys is
/// *promptness*: waking a laptop reconnects now rather than at the next
/// backoff step, and the reader can say "offline" instead of "still
/// downloading" for a body that is not coming. That distinction is the whole
/// reason those are separate absence states.
public final class Reachability {
    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "dev.postio.reachability")
    private var started = false

    public init() {}

    /// Report every change to `report`, until this object goes away.
    ///
    /// `report` is called on a background queue and may be called with the
    /// same answer twice — an interface changing while the path stays
    /// satisfied is a fresh callback saying nothing new. The boundary
    /// deliberately absorbs that: `setOffline` nudges only on a real
    /// transition, so no debounce is needed here and a second one would only
    /// add a place for the two to disagree.
    public func start(_ report: @escaping @Sendable (Bool) -> Void) {
        guard !started else { return }
        started = true
        monitor.pathUpdateHandler = { path in
            report(Self.isOffline(status: path.status))
        }
        monitor.start(queue: queue)
    }

    public func stop() {
        monitor.cancel()
    }

    /// Whether a path means "no connection right now".
    ///
    /// Separated from the watching so it can be asserted: `NWPathMonitor`
    /// needs a real interface and answers on its own schedule, and this is
    /// the part with a decision in it.
    static func isOffline(status: NWPath.Status) -> Bool {
        switch status {
        case .satisfied:
            return false
        case .unsatisfied:
            return true
        case .requiresConnection:
            // "Would be satisfied if something established it" — a
            // dial-on-demand VPN, typically. **Online**, because the engine is
            // exactly that something: it will try, and trying is what brings
            // the link up. Reporting offline would stop it making the attempt
            // that would have worked — every other mistake this watcher can
            // make costs a second, and this one would cost the connection.
            return false
        @unknown default:
            // A status this build has not heard of. Assuming a connection is
            // the safe default: the engine's own retry handles being wrong,
            // while claiming offline is a sentence the reader renders at
            // somebody.
            return false
        }
    }
}

File: crates/postio-ffi/tests/reachability.rs
//! Reachability pushed down from the platform (#663).
//!
//! The observation happens in Swift, where `NWPathMonitor` reads it in four
//! lines; Rust is told. Binding it here would mean `unsafe` in a crate that
//! forbids it, for a signal the platform already hands over.
//!
//! What is asserted is this side of that seam: the flag lands, and coming back
//! from offline is a *nudge* rather than only a flag change — the engine
//! reconnects with backoff on its own and works with no signal at all, so the
//! whole value of knowing is promptness.

use postio_ffi::{Session, SessionOptions};

fn session() -> std::sync::Arc<Session> {
    Session::open(SessionOptions::in_memory()).expect("an in-memory session")
}

#[test]
fn the_flag_lands_where_the_reader_can_see_it() {
    // `Absent::Offline` versus `Absent::Partial` is the whole reason this
    // setter exists (#591): a body that is not coming reads differently from
    // one still on its way.
    let session = session();
    assert!(
        !session.is_offline(),
        "a fresh session assumes a connection"
    );

    session.set_offline(true);
    assert!(session.is_offline());

    session.set_offline(false);
    assert!(!session.is_offline());
    session.shutdown();
}

#[test]
fn coming_back_online_asks_the_engine_to_try_again() {
    // The promptness half. Without it, waking a laptop waits out whatever
    // backoff step the engine had reached, which can be minutes.
    let session = session();
    session.set_offline(true);
    let before = session.reconnects_for_test();

    session.set_offline(false);
    assert_eq!(
        session.reconnects_for_test(),
        before + 1,
        "regaining the network prompts a reconnect"
    );
    session.shutdown();
}

#[test]
fn nothing_is_nudged_when_the_state_did_not_change() {
    // `NWPathMonitor` repeats itself: an interface changing while still
    // satisfied is a fresh callback with the same answer. Reconnecting on
    // every one of those would hammer the server on a flapping connection --
    // the opposite of what backoff is for.
    let session = session();
    session.set_offline(false);
    let before = session.reconnects_for_test();

    session.set_offline(false);
    session.set_offline(false);
    assert_eq!(
        session.reconnects_for_test(),
        before,
        "the same answer twice is not an event"
    );
    session.shutdown();
}

#[test]
fn going_offline_is_never_a_nudge() {
    let session = session();
    let before = session.reconnects_for_test();
    session.set_offline(true);
    assert_eq!(
        session.reconnects_for_test(),
        before,
        "losing the network is not a moment to try connecting"
    );
    session.shutdown();
}
